### PR TITLE
Three step written in CWL

### PR DESCRIPTION
**Pull request by danbills**
This is just a test expressing the classic WDL "Three step" in CWL.

This is post-re-org, so there are many commits included here and I don't know why.  Just look at the files changed.

### CODE REVIEW

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala [hunk 1]
```diff
@@ -1,51 +1,173 @@
...
+class ThreeStepSpec extends FlatSpec with Matchers {
+  val namespace = "threestep"
+  it should "print nicely " in {
+
+    val inlineJScriptRequirements = Option(Array(
+      Coproduct[Requirement](ShellCommandRequirement()),
+      Coproduct[Requirement](InlineJavascriptRequirement())))
+
+  /**
+    * PS
```
Comment (geoffjentry) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: It took me a lot longer than it should to realize what 'Ps' meant here. Not sure one should cater comments to my mental deficiencies, but .... :)

**Comment by geoffjentry** [comment]
👍

**Comment by mcovarr** [comment]
Yeah you might want to resort to the "one giant diffed commit" approach, it's becoming quite the rage. Whatever you merge should be squashed down to one commit first.

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala [hunk 2]
```diff
@@ -1,51 +1,173 @@
...
+    s => Coproduct[Argument](CommandLineBinding(valueFrom = Some(Coproduct[StringOrExpression](s)), shellQuote  = Some(false)))
+
+  val cgrepArgs = Array(
+    clb("grep"),
+    clb("$(inputs.pattern)"),
+    clb("$(inputs.file)"),
+    clb("|"),
+    clb("wc"),
+    clb("-l")
+  )
```
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: map the clb over a String array?

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala [hunk 3]
```diff
@@ -1,51 +1,173 @@
...
+  /**
+    * PS
+    */
+  val psOutputBinding = CommandOutputBinding(glob = Some(Coproduct[Glob](("ps-stdOut.txt"))))
+
+  val psOutputParameter =
+    CommandOutputParameter(
+      id = "ps-stdOut",
+      `type` = Option(Coproduct(CwlType.File)),
+      outputBinding = Some(psOutputBinding))
```
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: Option throughout

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala (hunk 1 above)
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: Yeah not sure why this was shouted, confused me too

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala (hunk 2 above)
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: Ooh or write as a legible String, split on white space, then map

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala [hunk 4]
```diff
@@ -1,51 +1,173 @@
...
+      inputs = _inputs,
+      outputs = _outputs,
+      steps = Coproduct[WorkflowSteps](Array(psWfStep, grepWfStep, wcWorkflowStep)))
+
+    val yaml = encodeWorkflow(threeStepWorkflow)
+
+    println(yaml)
+
+
+    true shouldBe true
```
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: I have a good feeling this is not going to be one of our flaky tests

Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: Srsly tho is this here to placate the framework or some other reason?

**Comment by danbills** [comment]
closing until squashed & WOMed

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala [hunk 5]
```diff
@@ -1,51 +1,173 @@
...
+
+  val _inputs = Coproduct[WorkflowInput](Array(workflowPatternInput))
+
+  val threeStepWorkflow =
+    new Workflow(
+      Some(CwlVersion.Version1),
+      `class` = "Workflow".narrow,
+      inputs = _inputs,
+      outputs = _outputs,
+      steps = Coproduct[WorkflowSteps](Array(psWfStep, grepWfStep, wcWorkflowStep)))
```
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: Probably a dumb question but it seems the dependency of the two downstream steps on ps is not expressed?

**Comment by mcovarr** [comment]
:+1:

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala (hunk 4 above)
Comment (danbills) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: my bad I just wanted to print this out over and over.  I'll remove the test

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala (hunk 5 above)
Comment (danbills) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: relative path that is resolved by [Schema Salad](http://www.commonwl.org/v1.0/SchemaSalad.html#Identifier_resolution) and passed in the input "ps/stdOut" `WorkFlowStepInput`'s source

**Comment by geoffjentry** [comment]
@danbills what's your plan of attack for schema salad? IIRC there's only a python version right now.

Also, do you think we need it? I believe the Bunny folks found it superfluous, is this an artifact of our CWL parser here?

**Comment by danbills** [comment]
@geoffjentry  I filed a bug #141 and added some explanation.

File: cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala (hunk 5 above)
Comment (mcovarr) [inline cwl/src/test/scala/wdl4s/cwl/ThreeStepSpec.scala]: Ah I totally missed that, thanks!

**Pull request closed without merging**